engine/src/scheme.rs

```diff
@@ -267,7 +267,7 @@ impl<'s> Scheme {
     pub fn add_function(
         &mut self,
         name: String,
-        function: Function,
+        function: impl FunctionDefinition + 'static,
```

Comment:
Since we move the Function Definition into heap allocation, should we just make the function signature `dyn Box<FunctionDefinition>`?

Reply:
Besides the fact that it would be a backward incompatible change, it would also mean leaking some implementation detail as the caller would now be forced to call `Box::new` itself. I think relying on monomorphism is fine here.